lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,423 @@
+#include "aws/s3/model/GetObjectRequest.h"
+#include "aws/s3/model/HeadObjectRequest.h"
+#include "aws/s3/model/PutObjectRequest.h"
+
+#include "llvm/Support/Process.h"
+#include "llvm/Support/YAMLTraits.h"
+
+#include "revng/Storage/Path.h"
+#include "revng/Storage/S3StorageClient.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/OnQuit.h"
+#include "revng/Support/PathList.h"
+#include "revng/Support/TemporaryFile.h"
+
+using FilenameMapType = llvm::StringMap<std::string>;
+
+template<>
+struct llvm::yaml::CustomMappingTraits<FilenameMapType> {
+  static void inputOne(IO &IO, llvm::StringRef Key, FilenameMapType &Value) {
+    IO.mapRequired(Key.str().c_str(), Value[Key]);
+  }
+
+  static void output(IO &IO, FilenameMapType &Value) {
+    for (auto &Pair : Value) {
+      IO.mapRequired(Pair.first().str().c_str(), Pair.second);
+    }
+  }
+};
+
+namespace revng::detail {
+
+namespace {
+
+using Aws::Utils::Logging::FormattedLogSystem;
+using Aws::Utils::Logging::LogLevel;
+
+Logger<> Logger("aws");
+
+class LoggerSystem : public FormattedLogSystem {
+private:
+  std::mutex Mutex;
+
+public:
+  LoggerSystem(LogLevel LogLevel) : FormattedLogSystem(LogLevel) {}
+  ~LoggerSystem() override = default;
+  void Flush() override { Logger.flush(); }
+
+  void ProcessFormattedStatement(Aws::String &&Statement) override {
+    // This function needs to be thread-safe, hence the lock_guard
+    // TODO: check if the mangled text is the result of aggressive flushed
+    //       and/or if buffering might alleviate this problem
+    std::lock_guard Guard(Mutex);
+    revng_log(Logger, Statement);
+  }
+};
+
+bool SDKIsInitialized = false;
+
+void initializeSDK() {
+  revng_assert(!SDKIsInitialized);
+
+  LogLevel Level(LogLevel::Info);
+  Aws::SDKOptions Options;
+
+  Options.loggingOptions.logLevel = Level;
+  Options.loggingOptions.logger_create_fn = [Level]() {
+    return std::make_shared<LoggerSystem>(Level);
+  };
+  Aws::InitAPI(Options);
+  OnQuit->add([Options = std::move(Options)] { Aws::ShutdownAPI(Options); });
+}
+
+} // namespace
+
+static llvm::StringRef consumeSplit(llvm::StringRef &Input, char SplitChar) {
+  std::size_t Pos = Input.find(SplitChar);
+  revng_assert(Pos != llvm::StringRef::npos);
+  llvm::StringRef Result = Input.slice(0, Pos);
+  Input = Input.drop_front(Pos + 1);
+  return Result;
+}
+
+template<typename T>
+static llvm::Error
+toError(const Aws::Utils::Outcome<T, Aws::S3::S3Error> &Request) {
+  revng_check(!Request.IsSuccess());
+  return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                 Request.GetError().GetMessage());
+}
+
+static std::string resolvePath(const llvm::StringRef Root,
+                               const llvm::StringRef Path) {
+  if (Root.empty())
+    return Path.str();
+  else if (Path.empty())
+    return Root.str();
+  else
+    return joinPath(Root, Path);
+}
+
+static std::string generateNewFilename(const llvm::StringRef Path) {
```

Comment:
Doc the pattern of the generated file name

Reply:
Added